qemu/tests/qemu_guest_agent.py

```diff
@@ -4168,15 +4184,50 @@ def _base_on_bg_check_resource_leak():
 
             if bg.is_alive():
                 qga_handles, qga_memory = _process_qga_resource()
-                # Generally the qga handles won't exceed 300, and the memory
-                # setting '3000' is an average value when the guest is not
-                # running redundant operations and apps.
-                if qga_handles > 300 or qga_memory > 3000:
-                    test.fail("QGA commands caused resource leak. handles are"
-                              " %s, memory is %s" % (qga_handles, qga_memory))
-                else:
-                    LOG_JOB.info("Current qga handles is %s, qga memory"
-                                 "is %s", qga_handles, qga_memory)
+                if qga_memory > 3000:
+                    qga_memory_previous = qga_memory
+                    qga_handles_previous = qga_handles
+                    qga_handles_latest, qga_memory_latest = _process_qga_resource()
+                    memory_high_peak = []
+                    memory_low_peak = []
+                    memory_list = []
+                    start_time = time.time()
+                    while int(time.time() - start_time) < 600:
+                        while qga_memory_latest >= qga_memory_previous:
+                            difference = qga_memory_latest - qga_memory_previous
+                            qga_memory_previous = qga_memory_latest
+                            time.sleep(5)
+                            qga_handles_latest, qga_memory_latest = _process_qga_resource()
+                            if qga_memory_latest < qga_memory_previous:
+                                memory_high_peak.append(qga_memory_previous)
+                            if int(time.time() - start_time) > 600:
+                                break
+
+                        while (int(time.time() - start_time) < 600 and
+                               qga_memory_latest <= qga_memory_previous):
+                            difference = qga_memory_latest - qga_memory_previous
+                            qga_memory_previous = qga_memory_latest
+                            time.sleep(5)
+                            qga_handles_latest, qga_memory_latest = _process_qga_resource()
+                            if qga_memory_latest > qga_memory_previous:
+                                memory_low_peak.append(qga_memory_previous)
+                    high_peak_value = _compare_peak_value(memory_high_peak)
+                    low_peak_value = _compare_peak_value(memory_low_peak)
+                    if high_peak_value == 0 and low_peak_value == 0:
+                        if (len(high_peak_value) < 3 or len(low_peak_value) < 3) and len(memory_list) < 6:
```

Comment:
I don't find anywhere calling `memory_list`, it would be 0 forever.

Reply:
Would you explain why we choose `< 3`?

Reply:
> I don't find anywhere calling `memory_list`, it would be 0 forever.

yeah, you're right, I missed that part.

Reply:
> Would you explain why we choose `< 3`?

first of all, that should be if (len(memory_high_peak) < 3 or len(memory_low_peak) < 3) and len(memory_list) < 6,
and  setting 3 is just mainly to ensure the basic conditions of compare.

Reply:
Still not clear why have this extra condition, let's discuss it offline.